### PR TITLE
[RPC] Add REST RPCType detection and error reply formatting

### PR DESCRIPTION
**Pull request by red-0ne**
# Summary:
This PR introduces support for `RPCType` detection for `REST` requests and adds a helper function for error reply formatting.

# Details:
This update replaces the partials package in the `poktroll` repository, making its functionality available to any consumer of the SDK.

## New Features:

1. `GetRPCType` Function:

* Detects `JSON-RPC` and `REST` requests.
* Accepts a `POKTHTTPRequest` and returns the corresponding `poktroll`'s `sharedtypes.RPCType`.

2. `FormatError` Helper Function:

Accepts an error, the original `POKTHTTPRequest`, and `RPCType`.
Returns a `POKTHTTPResponse` representing the error, formatted according to the `RPCType` and refined using the `POKTHTTPRequest`.

## Improvements:

Refactors the `POKTHTTPRequest/Response` codecs to return intermediary steps, reducing redundant (de)serializations by the SDK consumer.

This PR partially addresses ticket https://github.com/pokt-network/poktroll/issues/221

### CODE REVIEW

File: rpcdetect/jsonrpc.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// isJSONRPC checks if the given POKTHTTPRequest is a JSON-RPC request.
+func isJSONRPC(poktRequest *types.POKTHTTPRequest) bool {
+	contentType, ok := poktRequest.Header[contentTypeHeaderKey]
+	if !ok {
+		return false
+	}
+
+	if !slices.Contains(contentType.Values, "application/json") {
```
Comment (Olshansk) [inline rpcdetect/jsonrpc.go:59]: Move `applicaiton/json` into a local constant and reuse.

File: rpcdetect/jsonrpc.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	var err error
+	defaultJSONRPCErrorReplyBz, err = proto.Marshal(defaultJSONRPCErrorReply)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// jsonRPCPayload represents the JSON-RPC payload fields that are relevant for
```
Comment (Olshansk) [inline rpcdetect/jsonrpc.go:44]: Are we just using this for detection or more?

1. If it's purely for detection -> ename to `jsonRPCPayloadMeta`
2. If it's for more -> #PUC

Ditto elsewhere.

File: rpcdetect/rest.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+	var err error
+	defaultRESTErrorReplyBz, err = proto.Marshal(defaultRESTErrorReply)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// isREST checks if the given POKTHTTPRequest is a REST request.
+func isREST(poktRequest *types.POKTHTTPRequest) bool {
```
Comment (Olshansk) [inline rpcdetect/rest.go:45]: I feel like almost all the functions take a req as a param field.

Given that goland has receivers and an "object oriented approach", it's more appropriate than going down this function path.

File: rpcdetect/jsonrpc.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	payload, err := readJSONRPCPayload(poktRequestBz.BodyBz)
+	if err == nil {
+		requestId = payload.Id
+	}
+
+	errorReplyPayload := map[string]interface{}{
+		"jsonrpc": "2.0",
+		"id":      requestId,
+		"error": map[string]interface{}{
+			"code":    -32000,
```
Comment (Olshansk) [inline rpcdetect/jsonrpc.go:98]: Keep seeing this number everywhere

File: rpcdetect/jsonrpc.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	responseBodyBz, err := json.Marshal(errorReplyPayload)
+	if err != nil {
+		// If we fail to marshal the error reply payload, we return the default JSON-RPC
+		// error reply.
+		return defaultJSONRPCErrorReply, defaultJSONRPCErrorReplyBz
+	}
+
+	header := &types.Header{
+		Key:    contentTypeHeaderKey,
+		Values: []string{"application/json"},
```
Comment (Olshansk) [inline rpcdetect/jsonrpc.go:113]: Move to a constant?

**Review by Olshansk — CHANGES_REQUESTED** [review]
@red-0ne Great start!

Left a handful of comments that will likely be repeated, so stopping my review about halfway and will redo the review again afterwards.

File: rpcdetect/rpc_id.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package rpcdetect
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply rpcdetect/rpc_id.go:1]: Moved the whole package to `types` in order to have `isXXX` and `FormaError` as receivers.

File: rpcdetect/jsonrpc.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	defaultJSONRPCErrorReplyBz, err = proto.Marshal(defaultJSONRPCErrorReply)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// jsonRPCPayload represents the JSON-RPC payload fields that are relevant for
+// detecting JSON-RPC requests.
+type jsonRPCPayload struct {
+	Id      uint64 `json:"id"`
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply rpcdetect/jsonrpc.go:47]: Yes, struct with json tag want to be exported.
image

File: types/jsonrpc.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+// jsonRPCPayloadMeta represents the JSON-RPC payload fields that are relevant for
+// detecting JSON-RPC requests.
+type jsonRPCPayloadMeta struct {
+	Id      uint64 `json:"id"`
+	JSONRPC string `json:"jsonrpc"`
+	Method  string `json:"method"`
+}
+
+// init initializes the package level variables for the JSON-RPC error reply.
```
Comment (Olshansk) [inline types/jsonrpc.go:35]:
```suggestion
// init initializes the package level variables such as the JSON-RPC error reply.
```

File: types/rest.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+var (
+	// defaultRESTErrorReply is the default REST error reply to be sent if the
+	// formatRESTError function fails to format the appropriate one.
+	defaultRESTErrorReply   *POKTHTTPResponse
+	defaultRESTErrorReplyBz []byte
+)
+
+// init initializes the package level variables for the REST error reply.
```
Comment (Olshansk) [inline types/rest.go:18]:
```suggestion
// init initializes the package level variables such as the REST error reply.
```

File: types/rpc_id.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	contentTypeHeaderValueText = "text/plain"
+	defaultErrorMessage        = "Internal error"
+)
+
+var (
+	unsupportedRPCTypeErrorReply   *POKTHTTPResponse
+	unsupportedRPCTypeErrorReplyBz []byte
+)
+
+// init initializes the package level variables for the unsupported RPC type error reply.
```
Comment (Olshansk) [inline types/rpc_id.go:22]:
```suggestion
// init initializes the package level variables such as unsupported RPC type error reply.
```

File: types/rpc_id.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
+package types
```
Comment (Olshansk) [inline types/rpc_id.go:1]: Thoughts on renaming this file to `rpc_helpers`, `rpc_default` or `rpc_errors`?

Probably the last one is best.

**Review by Olshansk — APPROVED** [review]
Couple minor NITs but nothing blocking!

**Pull request merged**